pkg/controller/util.go

```diff
@@ -326,3 +329,297 @@ func addToMap(m1, m2 map[string]string) map[string]string {
+	if err != nil {
+		return nil, errors.Wrap(err, "clone target pod API create errored")
+	}
+	glog.V(Vuser).Infof("clone target pod \"%s/%s\" (image: %q) created\n", pod.Namespace, pod.Name, image)
+	return pod, nil
+}
+
+// return the clone target pod spec based on the target pvc.
+func MakeCloneTargetPodSpec(image, verbose, pullPolicy, cr, secret string, pvc *v1.PersistentVolumeClaim) *v1.Pod {
+	// target pod name contains the pvc name
+	podName := fmt.Sprintf("%s-", CLONER_TARGET_PODNAME)
+
+	pod := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: podName,
+			Annotations: map[string]string{
+				AnnCreatedBy: "yes",
+			},
+		},
+		Spec: v1.PodSpec{
+			Affinity: &v1.Affinity{
+				NodeAffinity: &v1.NodeAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+						NodeSelectorTerms: []v1.NodeSelectorTerm{
+							{
+								MatchExpressions: []v1.NodeSelectorRequirement{
+									{
+										Key:      "kubevirt.io/Host-Assisted-Cloning",
+										Operator: v1.NodeSelectorOpIn,
+										Values: []string{"true"},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			Containers: []v1.Container{
+				{
+					Command:		[]string{"/bin/sh"},
+					Name:            CLONER_TARGET_PODNAME,
+					Image:           image,
+					ImagePullPolicy: v1.PullPolicy(pullPolicy),
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      ImagePathName,
+							MountPath: CLONER_IMAGE_PATH,
+						},
+						{
+							Name:      SocketPathName,
+							MountPath: CLONER_SOCKET_PATH,
+						},
+					},
+					Args: []string{"-c", "/tmp/script.sh target" },
+				},
+			},
+			RestartPolicy: v1.RestartPolicyNever,
+			Volumes: []v1.Volume{
+				{
+					Name: ImagePathName,
+					VolumeSource: v1.VolumeSource{
+						PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc.Name,
+							ReadOnly:  false,
+						},
+					},
+				},
+				{
+					Name: SocketPathName,
+					VolumeSource: v1.VolumeSource{
+						HostPath: &v1.HostPathVolumeSource{
+							Path: CLONER_SOCKET_PATH,
+						},
+					},
+
+				},
+			},
+		},
+	}
+	return pod
+}
+
+// checkPVC verifies that the passed-in pvc is one we care about. Specifically, it must have the
+// CloneRequest annotation and it must not already be "in-progress". If the pvc passes these filters
+// then true is returned and the source and the target pods will be created. `AnnCloneRequest` indicates that the
+// pvc is targeted for the cloning job. `AnnCloneInProgress` indicates the  pvc is being processed.
+// Note: there is a race condition where the AnnCloneInProgress annotation is not seen in time and as
+// a result the source and the target pods can be created twice (or more, presumably). To reduce this window
+// a Get api call can be requested in order to get the latest copy of the pvc before verifying
+// its annotations.
```

Comment:
Can you close the race window completely by adding the AnnCloneInProgress annotation to the PVC before spawning the pods?  If the pods fail to start you can remove the annotation so that you will try again the next time.

Reply:
moved the annotation set before the creation of the pods. Still need to figure out how to remove the anno from the pvc if the pods fails. Right now, it's TODO

Reply:
You can check the status.phase of the pod in ProcessNextPodItem, and if failed/completed, act on the PVCs accordingly.

Reply:
:-) What I meant is that I need to see how to remove the anno from the map as we as calling a methond to remove the anno from the pvc.